tools/clang/lib/SPIRV/InitListHandler.h

```diff
@@ -0,0 +1,124 @@
+/// braces of even wired forms like float2x2 mat = {{1.}, {2., {{3.}}}, 4.};
+///
+/// Then we construct the final SPIR-V composite from the initializer list
+/// by traversing the type of the composte. This is done recursively in the
+/// depth first search manner, using the type of the composite as the root.
+///
+/// When we reach a scalar type, we will try to decode a scalar value from the
+/// front of the initializers queue. This may trigger composite extraction
+/// since the front of the queue may be a vector/etc. The leftover values after
+/// the extraction should be retained for the next decoding. Thus, we need
+/// another queue, scalars, to keep track of leftover unused scalar values.
+/// To adjust properly, when decoding values for a given type, we first try
+/// the scalar queue.
+///
+/// When we reach a composite type, we will try to construct a composite using
+/// the scalar values previously extracted and retained in the scalars queue.
+/// To optimize, if we have no leftover scalars and a value of the same type at
+/// the front of the initializers queue, we use the value as a whole.
+class InitListHandler {
+public:
+  /// Constructs an InitListHandler which uses the given emitter for normal
+  /// translation tasks. It will reuse the ModuleBuilder embedded in the given
+  /// emitter.
+  explicit InitListHandler(SPIRVEmitter &emitter);
+
+  /// Processes the given InitListExpr and returns the <result-id> for the final
+  /// SPIR-V value.
+  uint32_t process(const InitListExpr *expr);
+
+private:
+  /// \brief Wrapper method to create an error message and report it
+  /// in the diagnostic engine associated with this consumer.
+  template <unsigned N> DiagnosticBuilder emitError(const char (&message)[N]) {
+    const auto diagId =
+        diags.getCustomDiagID(clang::DiagnosticsEngine::Error, message);
+    return diags.Report(diagId);
+  }
+
+  /// Flattens the given InitListExpr and puts all non-InitListExpr AST nodes
+  /// into initializers.
+  void flatten(const InitListExpr *expr);
+
+  /// Decomposes the given Expr and puts all elements into the end of the
+  /// scalars queue.
+  void decompose(const Expr *expr);
+
+  /// Emits the necessary SPIR-V instructions to create a SPIR-V value of the
+  /// given type. The scalars and initializers queue will be used to fetch the
+  /// next value.
+  uint32_t createInitForType(QualType type);
+  uint32_t createInitForBuiltinType(QualType type);
+  uint32_t createInitForVectorType(QualType elemType, uint32_t count);
+  uint32_t createInitForMatrixType(QualType elemType, uint32_t rowCount,
+                                   uint32_t colCount);
+
+private:
+  SPIRVEmitter &theEmitter;
+  ModuleBuilder &theBuilder;
+  TypeTranslator &typeTranslator;
+  DiagnosticsEngine &diags;
+
+  /// A queue keeping track of unused AST nodes for initializers
+  std::deque<const Expr *> initializers;
+  /// A queue keeping track of previously extracted but unused scalars
```

Comment:
Add a comment about what this pair is. What does the uint32_t represent? IIUC it represents the result-id of the QualType which is the second member of the pair.